src/na/na_ofi.c

```diff
@@ -0,0 +1,2950 @@
+    struct fi_info *providers = NULL;
+    na_return_t ret = NA_SUCCESS;
+    int rc;
+
+    /* Only the first initiator queries OFI providers */
+    if (HG_UTIL_TRUE != hg_atomic_cas32(&nofi_gdata.nog_init_flag, 0, 1)) {
+        while(nofi_gdata.nog_providers == NULL)
+            usleep(1);
+        goto out;
+    }
+
+    /**
+     * Hints to query && filter providers.
+     *
+     * mode: operational mode, NA_OFI passes in context for communication calls.
+     * ep_type: reliable datagram (connection-less).
+     * caps: capabilities required.
+     * msg_order: guarantee that messages with same tag are ordered.
+     * (FI_ORDER_SAS - Send after send. If set, message send operations,
+     *  including tagged sends, are transmitted in the order submitted relative
+     *  to other message send. If not set, message sends may be transmitted out
+     *  of order from their submission).
+     */
+    hints = fi_allocinfo();
+    if (!hints) {
+        NA_LOG_ERROR("fi_allocinfo failed.\n");
+        ret = NA_NOMEM_ERROR;
+        goto out;
+    }
+
+    hints->mode               = FI_CONTEXT;
+    hints->ep_attr->type      = FI_EP_RDM;
+    hints->caps               = FI_TAGGED | FI_RMA;
+    hints->tx_attr->msg_order = FI_ORDER_SAS;
+    hints->rx_attr->msg_order = FI_ORDER_SAS;
+
+    hints->domain_attr->threading        = FI_THREAD_UNSPEC;
+    if (!strcmp(protocol_name, "sockets")) {
+        /* As "sockets" provider does not support manual progress and wait
+         * objects, set progress to auto for now. Note that the provider
+         * effectively creates a thread for internal progress in that case.
+         */
+        hints->domain_attr->control_progress = FI_PROGRESS_AUTO;
+        hints->domain_attr->data_progress    = FI_PROGRESS_AUTO;
+    } else {
+        hints->domain_attr->control_progress = FI_PROGRESS_MANUAL;
+        hints->domain_attr->data_progress    = FI_PROGRESS_MANUAL;
+    }
+    hints->domain_attr->av_type          = FI_AV_MAP;
+    hints->domain_attr->resource_mgmt    = FI_RM_ENABLED;
+
+    /**
+     * fi_getinfo:  returns information about fabric services.
+     * Pass NULL for name/service to list all providers supported with above
+     * requirement hints.
+     */
+    rc = fi_getinfo(NA_OFI_VERSION, /* OFI version requested */
+                    NULL,  /* Optional name or fabric to resolve */
+                    NULL,  /* Optional service name or port to request */
+                    0ULL,  /* Optional flag */
+                    hints, /* In: Hints to filter providers */
+                    &providers); /* Out: List of matching providers */
+    if (rc != 0) {
+        NA_LOG_ERROR("fi_getinfo failed, rc: %d(%s).", rc, fi_strerror(-rc));
+        fi_freeinfo(hints);
+        ret = NA_PROTOCOL_ERROR;
+        goto out;
+    }
+
+    assert(providers != NULL);
+    nofi_gdata.nog_refcount = 0;
+    hg_thread_mutex_init(&nofi_gdata.nog_mutex);
+    nofi_gdata.nog_providers = providers;
```

Comment:
the na_ofi_getinfo() with refcount protection, will only be called once atually.